Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go

```diff
@@ -534,10 +534,20 @@ func (e *Etcd) DeleteCollection(ctx api.Context, options *api.DeleteOptions, lis
 		}()
 	}
 	wg.Wait()
+
 	select {
 	case err := <-errs:
 		return nil, err
 	default:
+		anyLabels := listOptions.LabelSelector == nil || listOptions.LabelSelector.Empty()
+		anyFields := listOptions.FieldSelector == nil || listOptions.FieldSelector.Empty()
+
+		if anyLabels && anyFields && api.NamespaceValue(ctx) != "" {
+			// Try to delete the directory, which will only succeed if the directory is empty.
```

Comment:
It's up to the Kind to determine if it supports graceful deletion, which is only known when `Delete` calls `BeforeDelete` on the strategy associated with that Kind.

I think this code could be improved some more to no longer ignore the object returned here:
https://github.com/openshift/origin/pull/7344/files/2c74fc04f78d01bf6fc4fff7d2ff35f6f0139006#diff-51031c94aa64233e569e007fdbf96633R528

If a non-nil object is returned with no error, then at least one of the objects is still undergoing graceful termination, and we can avoid calling delete directory.